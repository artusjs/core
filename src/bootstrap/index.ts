export * from './decorator';
export * from './types';
