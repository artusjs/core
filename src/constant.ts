export const DEFAULT_LOADER = 'module';
export const LOADER_NAME_META = 'loader:name';

export const ArtusInjectPrefix = 'artus#';

export enum ArtusInjectEnum {
  Application = 'artus#application',
  Config = 'artus#config',
  DefaultContainerName = 'artus#default_container',
  Frameworks = 'artus#framework-config',
  LifecycleManager = 'artus#lifecycle-manager',
  Packages = 'artus#packages',
}

export enum ARTUS_DEFAULT_CONFIG_ENV {
  DEV = 'development',
  PROD = 'production',
  DEFAULT = 'default',
}

export enum ScanPolicy {
  NamedExport = 'named_export',
  DefaultExport = 'default_export',
  All = 'all',
}

export const ARTUS_EXCEPTION_DEFAULT_LOCALE = 'en';

export const ARTUS_SERVER_ENV = 'ARTUS_SERVER_ENV';

export const HOOK_NAME_META_PREFIX = 'hookName:';
export const HOOK_FILE_LOADER = 'appHook:fileLoader';

export const DEFAULT_EXCLUDES = [
  'test',
  'node_modules',
  '.*',
  'tsconfig*.json',
  '*.d.ts',
  'jest.config.*',
  'meta.json',
  'LICENSE',
  'pnpm-lock.yaml',
];
export const DEFAULT_MODULE_EXTENSIONS = ['.js', '.json', '.node'];

export const FRAMEWORK_PATTERN = 'framework.*';
export const PLUGIN_CONFIG_PATTERN = 'plugin.*';
export const PACKAGE_JSON = 'package.json';
export const PLUGIN_META_FILENAME = 'meta.json';
export const EXCEPTION_FILENAME = 'exception.json';

export const DEFAULT_LOADER_LIST_WITH_ORDER = [
  'exception',
  'exception-filter',
  'plugin-meta',
  'framework-config',
  'package-json',
  'module',
  'lifecycle-hook-unit',
  'config',
];

export const DEFAULT_CONFIG_DIR = 'src/config';
export const DEFAULT_MANIFEST_FILENAME = 'manifest.json';

export const SHOULD_OVERWRITE_VALUE = 'shouldOverwrite';
