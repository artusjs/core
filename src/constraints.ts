export const DEFAULT_LOADER = 'module';

export const enum ArtusInjectEnum {
  Application = 'artus#application',
  Trigger = 'artus#trigger',
  LifecycleManager = 'artus#lifecycle-manager',
  Config = 'artus#config',
  DefaultContainerName = 'artus#default_container',
}

export const ARTUS_EXCEPTION_DEFAULT_LOCALE = 'en';

export const ARTUS_SERVER_ENV = 'ARTUS_SERVER_ENV';

export enum ARTUS_DEFAULT_CONFIG_ENV {
  DEV = 'development',
  PROD = 'production',
  DEFAULT = 'default',
};

export const HOOK_CONSTRUCTOR_PARAMS = 'appHook:constructorParams';
export const HOOK_CONSTRUCTOR_PARAMS_APP = 'appHook:constructorParams:app';
export const HOOK_CONSTRUCTOR_PARAMS_CONTAINER = 'appHook:constructorParams:container';
export const HOOK_PARAMS_CONTEXT = 'appHook:constructorParams:context';
export const HOOK_NAME_META_PREFIX = 'hookName:';

export const DEFAULT_EXCLUDES = [
  'test',
  'node_modules',
  '.*',
  'tsconfig*.json',
  '*.d.ts',
  'jest.config.*',
  'package.json',
  'meta.*',
  'LICENSE'
];

export const PLUGIN_CONFIG_PATTERN = 'plugin.*';
export const CONFIG_PATTERN = 'config.*';
export const EXTENSION_PATTERN = 'app.*';
export const PLUGIN_META = ['meta.json', 'meta.yaml', 'meta.yml'];
export const PACKAGE_JSON = 'package.json';
export const EXCEPTION_FILE = 'artus-exception.yaml';

export const DEFAULT_LOADER_LIST_WITH_ORDER = [
  'exception',
  'plugin-config',
  'extension',
  'config',
  'plugin-meta',
  'module'
];
