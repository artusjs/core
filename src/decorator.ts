import {
  HOOK_CONSTRUCTOR_PARAMS,
  HOOK_CONSTRUCTOR_PARAMS_APP,
  HOOK_CONSTRUCTOR_PARAMS_CONTAINER,
  HOOK_NAME_META_PREFIX,
  HOOK_PARAMS_CONTEXT
} from './constraints';

export function ApplicationExtension(): ClassDecorator {
  return (_target: any) => {
    // Ready to remove?
  };
};

export function ApplicationHook(hookName?: string): PropertyDecorator {
  return (target: any, propertyKey: string | symbol) => {
    if (typeof propertyKey === 'symbol') {
      throw new Error(`hookName is not support symbol [${propertyKey.description}]`);
    }
    Reflect.defineMetadata(`${HOOK_NAME_META_PREFIX}${propertyKey}`, hookName ?? propertyKey, target.constructor);
  };
};

const WithApplicationExtensionParams = (tag: string): ParameterDecorator => {
  return (target: any, _propertyKey: string | symbol, parameterIndex: number) => {
    const paramsMd = Reflect.getOwnMetadata(HOOK_CONSTRUCTOR_PARAMS, target) ?? [];
    paramsMd[parameterIndex] = tag;
    if (_propertyKey) {
      Reflect.defineMetadata(HOOK_CONSTRUCTOR_PARAMS, paramsMd, target[_propertyKey]); // for proto
    } else {
      Reflect.defineMetadata(HOOK_CONSTRUCTOR_PARAMS, paramsMd, target); // for constructor
    }
  };
}

export function WithApplication(): ParameterDecorator {
  return WithApplicationExtensionParams(HOOK_CONSTRUCTOR_PARAMS_APP);
}

export function WithContainer(): ParameterDecorator {
  return WithApplicationExtensionParams(HOOK_CONSTRUCTOR_PARAMS_CONTAINER);
}

export function WithContext(): ParameterDecorator {
  return WithApplicationExtensionParams(HOOK_PARAMS_CONTEXT);
}

export * from './loader/decorator';
export * from './trigger/decorator';
