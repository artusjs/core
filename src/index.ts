

export {
  Inject,
  Injectable
} from '@artus/injection';

export {
  ArtusInjectEnum
} from './constraints';
export * from './loader';
export * from './lifecycle';
export * from './exception';
export * from './plugin';
export * from './application';
export * from './scanner';

import Trigger from './trigger';
export { Trigger };

export type {
  Manifest
} from './loader/types';
