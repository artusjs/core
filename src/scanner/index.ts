export * from './impl';
export * from './types';

