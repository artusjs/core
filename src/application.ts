import 'reflect-metadata';
import { Container } from '@artus/injection';
import { ArtusInjectEnum } from './constant';
import { ArtusStdError } from './exception';
import { HookFunction, LifecycleManager } from './lifecycle';
import { LoaderFactory, Manifest } from './loader';
import { mergeConfig } from './loader/utils/merge';
import { Application, ApplicationInitOptions } from './types';
import ConfigurationHandler from './configuration';
import { Logger, LoggerType } from './logger';

export class ArtusApplication implements Application {
  public manifest?: Manifest;
  public container: Container;

  constructor(opts?: ApplicationInitOptions) {
    this.container = new Container(opts?.containerName ?? ArtusInjectEnum.DefaultContainerName);

    if (opts?.env) {
      const envList = [].concat(opts.env);
      this.container.set({ id: ArtusInjectEnum.EnvList, value: envList });
    }
    this.loadDefaultClass();
    this.addLoaderListener();

    process.on('SIGINT', () => this.close(true));
    process.on('SIGTERM', () => this.close(true));
  }

  get config(): Record<string, any> {
    return this.container.get(ArtusInjectEnum.Config);
  }

  get configurationHandler(): ConfigurationHandler {
    return this.container.get(ConfigurationHandler);
  }

  get lifecycleManager(): LifecycleManager {
    return this.container.get(LifecycleManager);
  }

  get loaderFactory(): LoaderFactory {
    return this.container.get(LoaderFactory);
  }

  get logger(): LoggerType {
    return this.container.get(Logger);
  }

  loadDefaultClass() {
    // load Artus default clazz
    this.container.set({ id: Container, value: this.container });
    this.container.set({ id: ArtusInjectEnum.Application, value: this });
    this.container.set({ id: ArtusInjectEnum.Config, value: {} });

    this.container.set({ type: ConfigurationHandler });
    this.container.set({ type: LoaderFactory });
    this.container.set({ type: LifecycleManager });
    this.container.set({ type: Logger });
  }

  async load(manifest: Manifest, root: string = process.cwd()) {
    // Load user manifest
    this.manifest = manifest;

    await this.loaderFactory.loadManifest(manifest, root);

    await this.lifecycleManager.emitHook('didLoad');

    return this;
  }

  async run() {
    await this.lifecycleManager.emitHook('willReady'); // 通知协议实现层启动服务器
    await this.lifecycleManager.emitHook('didReady');
  }

  registerHook(hookName: string, hookFn: HookFunction) {
    this.lifecycleManager.registerHook(hookName, hookFn);
  }

  async close(exit = false) {
    try {
      // reverse emitHook to avoid plugin closed before app hook
      await this.lifecycleManager.emitHook('beforeClose', null, true);
    } catch (e) {
      throw new Error(e);
    }
    if (exit) {
      process.exit(0);
    }
  }

  throwException(code: string): void {
    throw new ArtusStdError(code);
  }

  createException(code: string): ArtusStdError {
    return new ArtusStdError(code);
  }

  protected addLoaderListener() {
    this.loaderFactory
      .addLoaderListener('config', {
        before: () => this.lifecycleManager.emitHook('configWillLoad'),
        after: () => {
          this.updateConfig();
          return this.lifecycleManager.emitHook('configDidLoad');
        },
      });
  }

  protected updateConfig() {
    const oldConfig = this.container.get(ArtusInjectEnum.Config, { noThrow: true }) ?? {};
    const newConfig = this.configurationHandler.getMergedConfig() ?? {};
    this.container.set({
      id: ArtusInjectEnum.Config,
      value: mergeConfig(oldConfig, newConfig),
    });
  }
}
