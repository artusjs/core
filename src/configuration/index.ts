import { Container, Inject, Injectable } from '@artus/injection';
import { ArtusInjectEnum, ARTUS_DEFAULT_CONFIG_ENV, ARTUS_SERVER_ENV } from '../constant';
import { ManifestItem } from '../loader';
import { mergeConfig } from '../loader/utils/merge';
import compatibleRequire from '../utils/compatible_require';

export type ConfigObject = Record<string, any>;
export type PackageObject = ConfigObject;

@Injectable()
export default class ConfigurationHandler {
  static getEnvFromFilename(filename: string): string {
    let [_, env, extname] = filename.split('.');
    if (!extname) {
      env = ARTUS_DEFAULT_CONFIG_ENV.DEFAULT;
    }
    return env;
  }

  public configStore: Record<string, ConfigObject> = {};

  @Inject()
  private container: Container;

  getMergedConfig(): ConfigObject {
    return this.mergeConfigByStore(this.configStore);
  }

  mergeConfigByStore(store: Record<string, ConfigObject>): ConfigObject {
    let envList: string[] = this.container.get(ArtusInjectEnum.EnvList, { noThrow: true });
    if (!envList) {
      envList = process.env[ARTUS_SERVER_ENV] ? [process.env[ARTUS_SERVER_ENV]] : [ARTUS_DEFAULT_CONFIG_ENV.DEV];
    }
    const defaultConfig = store[ARTUS_DEFAULT_CONFIG_ENV.DEFAULT] ?? {};
    const envConfigList = envList.map(currentEnv => (store[currentEnv] ?? {}));
    return mergeConfig(defaultConfig, ...envConfigList);
  }

  clearStore(): void {
    this.configStore = {};
  }

  setConfig(env: string, config: ConfigObject) {
    const storedConfig = this.configStore[env] ?? {};
    this.configStore[env] = mergeConfig(storedConfig, config);
  }

  async setConfigByFile(fileItem: ManifestItem) {
    const configContent: ConfigObject = await compatibleRequire(fileItem.path + fileItem.extname);
    if (configContent) {
      const env = ConfigurationHandler.getEnvFromFilename(fileItem.filename);
      this.setConfig(env, configContent);
    }
  }
}
