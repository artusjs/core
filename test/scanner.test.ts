import 'reflect-metadata';
import { ArtusScanner } from '../src/scanner';
import path from 'path';
import { ScanPolicy } from '../src';

describe('test/scanner.test.ts', () => {
  it('should be scan application', async () => {
    const scanner = new ArtusScanner({ needWriteFile: false, extensions: ['.ts'] });
    const manifest = await scanner.scan(path.resolve(__dirname, './fixtures/app_koa_with_ts'));
    expect(manifest.version).toBe('2');
    expect(manifest).toMatchSnapshot();

    // scan with relative root
    const relativeRoot = path.relative(process.cwd(), __dirname);
    const relativeRootManifest = await scanner.scan(path.join(relativeRoot, './fixtures/app_koa_with_ts'));
    expect(relativeRootManifest).toStrictEqual(manifest);
  });

  it('should scan application with all injectable class', async () => {
    const scanner = new ArtusScanner({ needWriteFile: false, extensions: ['.ts'] });
    const manifest = await scanner.scan(path.resolve(__dirname, './fixtures/named_export'));
    expect(manifest?.refMap?._app?.items).toBeDefined();
    expect(manifest.refMap._app.items.length).toBe(4);
  });

  it('should scan application with named export class', async () => {
    const scanner = new ArtusScanner({ needWriteFile: false, extensions: ['.ts'], policy: ScanPolicy.NamedExport });
    const manifest = await scanner.scan(path.resolve(__dirname, './fixtures/named_export'));
    expect(manifest?.refMap?._app?.items).toBeDefined();
    expect(manifest?.refMap?._app?.items.length).toBe(4);
  });

  it('should scan application with default export class', async () => {
    const scanner = new ArtusScanner({ needWriteFile: false, extensions: ['.ts'], policy: ScanPolicy.DefaultExport });
    const manifest = await scanner.scan(path.resolve(__dirname, './fixtures/named_export'));
    expect(manifest?.refMap?._app?.items).toBeDefined();
    expect(manifest?.refMap?._app?.items.length).toBe(2);
  });

  it('should not throw when scan application without configdir', async () => {
    const scanner = new ArtusScanner({ needWriteFile: false, extensions: ['.ts'] });
    const manifest = await scanner.scan(path.resolve(__dirname, './fixtures/app_without_config'));
    expect(manifest?.refMap?._app?.items?.find(item => item.loader === 'config')).toBeUndefined();
  });
});

