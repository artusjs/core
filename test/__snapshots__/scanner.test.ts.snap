// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test/scanner.test.ts should be scan application 1`] = `
{
  "extraPluginConfig": {},
  "refMap": {
    "@artus/injection": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "relativedPath": "../../../node_modules/@artus/injection/lib",
    },
    "_app": {
      "items": [
        {
          "extname": ".ts",
          "filename": "config.default.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "src/config/config.default",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".ts",
          "filename": "config.dev.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "src/config/config.dev",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".ts",
          "filename": "plugin.default.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "src/config/plugin.default",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".ts",
          "filename": "plugin.dev.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "src/config/plugin.dev",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".ts",
          "filename": "hello.ts",
          "loader": "module",
          "loaderState": {
            "exportNames": [
              "default",
            ],
          },
          "path": "src/controllers/hello",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".json",
          "filename": "exception.json",
          "loader": "exception",
          "loaderState": {
            "exportNames": [],
          },
          "path": "src/exception",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".ts",
          "filename": "default.ts",
          "loader": "exception-filter",
          "loaderState": {
            "exportNames": [
              "MockExceptionFilter",
            ],
          },
          "path": "src/filter/default",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".ts",
          "filename": "koa_app.ts",
          "loader": "module",
          "loaderState": {
            "exportNames": [
              "default",
            ],
          },
          "path": "src/koa_app",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".ts",
          "filename": "lifecycle.ts",
          "loader": "lifecycle-hook-unit",
          "loaderState": {
            "exportNames": [
              "default",
            ],
          },
          "path": "src/lifecycle",
          "source": "app",
          "unitName": "_app",
        },
        {
          "extname": ".ts",
          "filename": "hello.ts",
          "loader": "module",
          "loaderState": {
            "exportNames": [
              "default",
            ],
          },
          "path": "src/services/hello",
          "source": "app",
          "unitName": "_app",
        },
      ],
      "packageVersion": undefined,
      "pluginConfig": {
        "default": {
          "mysql": {
            "enable": false,
            "refName": "src/mysql_plugin",
          },
          "redis": {
            "enable": true,
            "refName": "src/redis_plugin",
          },
          "testDuplicate": {
            "enable": false,
            "refName": "@artus/injection",
          },
        },
        "dev": {
          "testDuplicate": {
            "enable": true,
            "refName": "src/test_duplicate_plugin",
          },
        },
      },
      "relativedPath": "",
    },
    "src/redis_plugin": {
      "items": [
        {
          "extname": ".ts",
          "filename": "app.ts",
          "loader": "lifecycle-hook-unit",
          "loaderState": {
            "exportNames": [
              "default",
            ],
          },
          "path": "src/redis_plugin/app",
          "source": "plugin",
          "unitName": "redis",
        },
      ],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "exclude": [
          "not_to_be_scanned_dir",
          "not_to_be_scanned_file.ts",
        ],
        "name": "redis",
      },
      "relativedPath": "src/redis_plugin",
    },
    "src/test_duplicate_plugin": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "name": "testDuplicate",
      },
      "relativedPath": "src/test_duplicate_plugin",
    },
  },
  "version": "2",
}
`;

exports[`test/scanner.test.ts should scan application with nesting preset a which defined in options 1`] = `
{
  "extraPluginConfig": {
    "preset_a": {
      "enable": true,
      "refName": "../plugins/preset_a",
    },
  },
  "refMap": {
    "../plugins/plugin_a": {
      "items": [],
      "packageVersion": "0.0.1",
      "pluginConfig": {},
      "pluginMetadata": {
        "dependencies": [
          {
            "name": "plugin-b",
          },
          {
            "name": "plugin-c",
            "optional": true,
          },
        ],
        "name": "plugin-a",
      },
      "relativedPath": "../plugins/plugin_a",
    },
    "../plugins/plugin_b": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "dependencies": [
          {
            "name": "plugin-c",
          },
        ],
        "name": "plugin-b",
      },
      "relativedPath": "../plugins/plugin_b",
    },
    "../plugins/plugin_d": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "dependencies": [
          {
            "name": "plugin-c",
            "optional": true,
          },
        ],
        "name": "plugin-d",
      },
      "relativedPath": "../plugins/plugin_d",
    },
    "../plugins/plugin_with_entry_a": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "name": "plugin-with-entry-a",
      },
      "relativedPath": "../plugins/plugin_with_entry_a/mock_lib",
    },
    "../plugins/plugin_with_entry_b": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "name": "plugin-with-entry-b",
      },
      "relativedPath": "../plugins/plugin_with_entry_b/mock_lib",
    },
    "../plugins/preset_a": {
      "items": [
        {
          "extname": ".ts",
          "filename": "plugin.default.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "../plugins/preset_a/config/plugin.default",
          "source": "plugin",
          "unitName": "preset_a",
        },
      ],
      "packageVersion": undefined,
      "pluginConfig": {
        "default": {
          "a": {
            "enable": false,
          },
          "plugin-with-entry-a": {
            "enable": true,
            "refName": "../plugins/plugin_with_entry_a",
          },
          "preset_b": {
            "enable": true,
            "refName": "../plugins/preset_b",
          },
          "preset_c": {
            "enable": true,
            "refName": "../plugins/preset_c",
          },
        },
      },
      "pluginMetadata": {
        "name": "preset_a",
      },
      "relativedPath": "../plugins/preset_a",
    },
    "../plugins/preset_b": {
      "items": [
        {
          "extname": ".ts",
          "filename": "plugin.default.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "../plugins/preset_b/config/plugin.default",
          "source": "plugin",
          "unitName": "preset_b",
        },
      ],
      "packageVersion": undefined,
      "pluginConfig": {
        "default": {
          "a": {
            "enable": true,
            "refName": "../plugins/plugin_a",
          },
          "b": {
            "enable": true,
            "refName": "../plugins/plugin_b",
          },
          "plugin-with-entry-b": {
            "enable": true,
            "refName": "../plugins/plugin_with_entry_b",
          },
        },
      },
      "pluginMetadata": {
        "name": "preset_b",
      },
      "relativedPath": "../plugins/preset_b",
    },
    "../plugins/preset_c": {
      "items": [
        {
          "extname": ".ts",
          "filename": "plugin.default.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "../plugins/preset_c/config/plugin.default",
          "source": "plugin",
          "unitName": "preset_c",
        },
      ],
      "packageVersion": undefined,
      "pluginConfig": {
        "default": {
          "b": {
            "enable": false,
            "refName": "../plugins/plugin_b",
          },
          "c": {
            "enable": false,
            "refName": "../plugins/plugin_c",
          },
          "d": {
            "enable": true,
            "refName": "../plugins/plugin_d",
          },
          "plugin-with-entry-c": {
            "enable": false,
            "refName": "../plugins/plugin_with_entry_c",
          },
        },
      },
      "pluginMetadata": {
        "name": "preset_c",
      },
      "relativedPath": "../plugins/preset_c",
    },
    "_app": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "relativedPath": "",
    },
  },
  "version": "2",
}
`;

exports[`test/scanner.test.ts should scan application with single preset b which defined in config 1`] = `
{
  "extraPluginConfig": {},
  "refMap": {
    "../plugins/plugin_a": {
      "items": [],
      "packageVersion": "0.0.1",
      "pluginConfig": {},
      "pluginMetadata": {
        "dependencies": [
          {
            "name": "plugin-b",
          },
          {
            "name": "plugin-c",
            "optional": true,
          },
        ],
        "name": "plugin-a",
      },
      "relativedPath": "../plugins/plugin_a",
    },
    "../plugins/plugin_b": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "dependencies": [
          {
            "name": "plugin-c",
          },
        ],
        "name": "plugin-b",
      },
      "relativedPath": "../plugins/plugin_b",
    },
    "../plugins/plugin_with_entry_b": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "name": "plugin-with-entry-b",
      },
      "relativedPath": "../plugins/plugin_with_entry_b/mock_lib",
    },
    "../plugins/preset_b": {
      "items": [
        {
          "extname": ".ts",
          "filename": "plugin.default.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "../plugins/preset_b/config/plugin.default",
          "source": "plugin",
          "unitName": "preset_b",
        },
      ],
      "packageVersion": undefined,
      "pluginConfig": {
        "default": {
          "a": {
            "enable": true,
            "refName": "../plugins/plugin_a",
          },
          "b": {
            "enable": true,
            "refName": "../plugins/plugin_b",
          },
          "plugin-with-entry-b": {
            "enable": true,
            "refName": "../plugins/plugin_with_entry_b",
          },
        },
      },
      "pluginMetadata": {
        "name": "preset_b",
      },
      "relativedPath": "../plugins/preset_b",
    },
    "_app": {
      "items": [
        {
          "extname": ".ts",
          "filename": "plugin.default.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "config/plugin.default",
          "source": "app",
          "unitName": "_app",
        },
      ],
      "packageVersion": undefined,
      "pluginConfig": {
        "default": {
          "preset_b": {
            "enable": true,
            "refName": "../plugins/preset_b",
          },
        },
      },
      "relativedPath": "",
    },
  },
  "version": "2",
}
`;

exports[`test/scanner.test.ts should scan application with single preset c which defined in options 1`] = `
{
  "extraPluginConfig": {
    "preset_c": {
      "enable": true,
      "refName": "../plugins/preset_c",
    },
  },
  "refMap": {
    "../plugins/plugin_d": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "pluginMetadata": {
        "dependencies": [
          {
            "name": "plugin-c",
            "optional": true,
          },
        ],
        "name": "plugin-d",
      },
      "relativedPath": "../plugins/plugin_d",
    },
    "../plugins/preset_c": {
      "items": [
        {
          "extname": ".ts",
          "filename": "plugin.default.ts",
          "loader": "config",
          "loaderState": {
            "exportNames": [],
          },
          "path": "../plugins/preset_c/config/plugin.default",
          "source": "plugin",
          "unitName": "preset_c",
        },
      ],
      "packageVersion": undefined,
      "pluginConfig": {
        "default": {
          "b": {
            "enable": false,
            "refName": "../plugins/plugin_b",
          },
          "c": {
            "enable": false,
            "refName": "../plugins/plugin_c",
          },
          "d": {
            "enable": true,
            "refName": "../plugins/plugin_d",
          },
          "plugin-with-entry-c": {
            "enable": false,
            "refName": "../plugins/plugin_with_entry_c",
          },
        },
      },
      "pluginMetadata": {
        "name": "preset_c",
      },
      "relativedPath": "../plugins/preset_c",
    },
    "_app": {
      "items": [],
      "packageVersion": undefined,
      "pluginConfig": {},
      "relativedPath": "",
    },
  },
  "version": "2",
}
`;
